src/main/java/org/tarantool/RoundRobinSocketProviderImpl.java

```diff
@@ -2,123 +2,106 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 /**
  * Basic reconnection strategy that changes addresses in a round-robin fashion.
  * To be used with {@link TarantoolClientImpl}.
  */
-public class RoundRobinSocketProviderImpl implements SocketChannelProvider {
-    /** Timeout to establish socket connection with an individual server. */
-    private int timeout; // 0 is infinite.
-    /** Limit of retries. */
-    private int retriesLimit = -1; // No-limit.
-    /** Server addresses as configured. */
-    private final String[] addrs;
-    /** Socket addresses. */
-    private final InetSocketAddress[] sockAddrs;
-    /** Current position within {@link #sockAddrs} array. */
-    private int pos;
+public class RoundRobinSocketProviderImpl extends BaseSocketChannelProvider implements RefreshableSocketProvider {
 
     /**
-     * Constructs an instance.
-     *
-     * @param addrs Array of addresses in a form of [host]:[port].
+     * Socket addresses.
      */
-    public RoundRobinSocketProviderImpl(String... addrs) {
-        if (addrs == null || addrs.length == 0)
-            throw new IllegalArgumentException("addrs is null or empty.");
-
-        this.addrs = Arrays.copyOf(addrs, addrs.length);
+    private final List<InetSocketAddress> socketAddresses = new ArrayList<>();
 
-        sockAddrs = new InetSocketAddress[this.addrs.length];
-
-        for (int i = 0; i < this.addrs.length; i++) {
-            sockAddrs[i] = parseAddress(this.addrs[i]);
-        }
-    }
+    /**
+     * Current position within {@link #socketAddresses} array.
+     */
+    private AtomicInteger currentPosition = new AtomicInteger(-1);
 
     /**
-     * @return Configured addresses in a form of [host]:[port].
+     * Lock
      */
-    public String[] getAddresses() {
-        return this.addrs;
-    }
+    private ReadWriteLock addressListLock = new ReentrantReadWriteLock();
 
     /**
-     * Sets maximum amount of time to wait for a socket connection establishment
-     * with an individual server.
-     *
-     * Zero means infinite timeout.
+     * Constructs an instance.
      *
-     * @param timeout Timeout value, ms.
-     * @return {@code this}.
-     * @throws IllegalArgumentException If timeout is negative.
+     * @param addresses Array of addresses in a form of [host]:[port].
      */
-    public RoundRobinSocketProviderImpl setTimeout(int timeout) {
-        if (timeout < 0)
-            throw new IllegalArgumentException("timeout is negative.");
-
-        this.timeout = timeout;
+    public RoundRobinSocketProviderImpl(String... addresses) {
+        if (addresses.length == 0) {
+            throw new IllegalArgumentException("Addresses list must contain at least one address.");
```

Comment:
I need to figure out how we can support this feature w/o extra effort. Maybe we can init a client in some different state rather than RECONNECTING and wait for refreshing nodes list from the discoverer. If the process takes more than init timeout then we close the current client.